design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,613 @@
+Authentication factor (1)
+  |- Cookie
+  |- Kerberos
+  |- Username Password Form
+Authentication factor (2)
+  |- Cookie
+  |- Kerberos
+  |- OTP Form
+  |- FIDO-2
+```
+Potentially we could also introduce another type of authentication execution, we could call it "Authenticated on", to simplify the authenticators that bypass the authentication factors. We could then have:
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authenticated on
+  |- Cookie
+  |- Kerberos
+Authentication factor (1)
+  |- Username Password Form
+Authentication factor (2)
+  |- OTP Form
+  |- FIDO-2
+```
+
+#### Authentication categories
+
+Authentication categories allow the link between the actual execution within the flow and the
+credentials of a certain type.
+
+For example, currently there is the OTPFormAuthenticator for the browser flow and ValidateOTP for the direct grant flow.
+However, both these authentication executions will accept input from the same OTP credentials. This means that schematically, we would have:
+```
+-----------------------         ----------------------
+| OTP Credential 1    |---------|                    |    ------------------------
+-----------------------         |                    |----| OTPFormAuthenticator |
+-----------------------         | OTP Authentication |    ------------------------
+| OTP Credential 2    |---------|      Category      |    ---------------
+-----------------------         |                    |----| ValidateOTP |
+-----------------------         |                    |    ---------------
+| OTP Credential 3    |---------|                    |
+-----------------------         ----------------------
+```
+
+This schema serves as a base for all credentials and allows them to be
+used for the appropriate authentication executions. This link allows Keycloak to present
+alternative credentials for an execution during login.
+
+To simplify selection in the GUI, it would also be useful to do some refactoring to ensure
+that the execution code for the `Browser` / `Registration Flow` / `Direct Grant Flow` /
+`Reset Credentials` / `Client Authentication` Bindings can only be selected as executions
+for their binding categories. This already seems to be the case for `Client Authentication`,
+and would simplify the selection of authentication executions within a authentication
+factor.
+
+It may also allow a simplification of the naming when selecting an authentication
+category within an authentication factor. For example, we could just
+have OTP instead of OTP Form (for `Browser` binding) and OTP (for `Direct Grant`)
+
+In addition to its function, an authentication category will also have a set of
+data/metadata that describes it. This can be a general field that will be found
+in all authenticator categories, for example description information (internationalised),
+an icon or something more functional like data for OIDC's amr claims (see section on
```

Comment:
I think it will be good if values of "amr" claims are configurable as well. For example some admin may want something like "amr" with value "otp" in the token, but some other admin will prefer to have something like "urn:something:otp". So maybe we need some configurable option on authentication category (or authentication execution) for this? Current userSession contains list of used authentication executions during the authentication. Not sure if this is sufficient information or something more needs to be added to userSession to being able to properly add the "amr" values into the token..

Reply:
The amr values are defined in [RFC 8176](https://tools.ietf.org/html/rfc8176), and I thought this should be sufficient. I was initially intending to have the authentication category statically defined in the code and in resource files, but if we are moving towards more configurable values (for this and for OTP policies for example, although this is still being discussed), then it will also require a table in the database.

I don't see why not having this configurable, for the purpose of greater compatibility, with the default values being those from the RFC.

Reply:
+1 for having it configurable. It will be probably good to have the protocol mapper, which will somehow support the mapping of "internal" execution to the `amr` value. As @stianst pointed, we can have different clients wanting different values, but I guess that usual case will be more clients sharing same value. We can probably introduce some new default "Client scope" and add the protocol mapper for adding `amr` as part of this client scope. That will allow more clients share the same `amr` values, but will also other clients to use specific values.

IMO this is something, which can be surely added in some later stage - will be nice to avoid have one big PR containing all the changes, but ideally divide it as much as possible to more PRs for easier review :)